File: cpp/4tank/src/TankController.cpp
//
// title: TankController.cpp
// by: Brian Kim
// description: class definition of the tank controller
//

#include "TankController.hpp"
#include "Constants.h"

TankController::TankController()
{
}


TankController::TankController( TankInfo info )
{
  // init drain
  _drain = Drain( info.drain_radius, info.drain_height, 1.0 );

  // calculate the openness of the tank's valve
  double tank_volume = PI * info.tank_radius * info.tank_radius;
  double tank_valve = _drain.getVolume() / tank_volume;

  // init tank
  _tank = Tank( info.tank_radius, info.tank_height, tank_valve );

  // init water 
  _waterVolume = info.init_water_volume;
  _waterIn = 0.0;
  _waterOut = 0.0;

  // init drain water volume
  if (_waterVolume)
    _drainWaterVolume = _waterVolume > _drain.getVolume() ? _drain.getVolume() : _waterVolume;
  else
    _drainWaterVolume = 0.0;

  // init state
  _state = _waterVolume ? TankStateStatic : TankStateEmpty;
}


double TankController::getWaterHeight() 
{
  double tank_radius = _tank.getRadius();
  
  return (_waterVolume - _drainWaterVolume) / (PI * tank_radius * tank_radius);
}

bool TankController::isDraining() { return _waterOut - _waterIn > 0.001; }
bool TankController::isEmpty()    { return _waterVolume < 0.00001; }
bool TankController::isFilling()  { return _waterOut - _waterIn < 0.001; }
bool TankController::isFull()     { return getTotalVolume() - _waterVolume < 0.000001; }
bool TankController::isOverflow() { return (isFull() && isFilling()) || (_waterVolume > getTotalVolume()); }
bool TankController::isStatic()   { return !isFull() && !isFilling(); }

void TankController::updateState() 
{   
  if      (isEmpty())    _state = TankStateEmpty;
  else if (isOverflow()) _state = TankStateOverflow;
  else if (isFull())     _state = TankStateFull;
  else if (isFilling())  _state = TankStateFilling;
  else if (isDraining()) _state = TankStateDraining;
  else                   _state = TankStateStatic;
}

void TankController::step()
{
  // dummy variables
  double drain_volume = _drain.getVolume();

  // temporary water volume 
  double t_waterVolume = _waterVolume + _waterIn;

  // only set water volume if tank isn't full
  if (_state != TankStateFull) _waterVolume = t_waterVolume;

  // only update water out if there is water
  if (_waterVolume)
  {
    _waterOut =  drain_volume * _drain.getValve();

    _waterVolume -= _waterOut;
    _drainWaterVolume -= _waterVolume > drain_volume ? 0 : _waterOut;

    // make sure certain values aren't < 0
    _waterVolume = _waterVolume > 0 ? _waterVolume : 0;
    _drainWaterVolume = _drainWaterVolume > 0 ? _drainWaterVolume : 0;
  }

  // update the state
  updateState();
}

File: py/4tank/model/TankController.py
##
# title: TankController.py
# by: Brian Kim
# description: a controller of the tank
#

from Tank import Tank
from Drain import Drain

# inner tank state enumeration
class TankState:
  Empty, Draining, Static, Filling, Full, Overflow = range(6)
  translator = { Empty : "Empty", Draining : "Draining", Static : "Static",
                 Filling : "Filling", Full : "Full", Overflow : "Overflow" }

class TankController():
  # 
  # accessor methods
  #
  def capacity( self ):
    return float(self._tank.volume() + self._drain.volume())

  def drain_volume( self ):
    return float(self._drain.volume())
    
  def drain_water_volume( self ):
    return float(self._drain_water_volume)

  def state( self ):
    return self._state

  def tank_volume( self ):
    return float(self._tank.volume())

  def valve_ratio( self, x=None ):
    if x == None:
      return float(self._drain.valve(x))
    else:
      self._drain.valve(x)

  def water_height( self ):
    return float(self._water_volume / self.capacity())
    
  def water_in( self, x=None ):
    if x == None:
      return float(self._water_in)
    else:
      self._water_in = x

  def water_out( self ):
    return float(self._water_out)

  def water_spill( self ):
    return self._water_spill

  def water_volume( self ):
    return float(self._water_volume)

  #
  # constructor
  # 
  def __init__( self, tank_radius=3, tank_height=5, drain_radius=1, 
                      drain_height=1, valve_ratio=0.5, water_height=0.15 ):
    # solid properties
    self._tank  =  Tank( tank_radius, tank_height, float(drain_radius)/tank_radius )
    self._drain = Drain( drain_radius, drain_height, valve_ratio )

    # fluid properties
    self._water_volume = water_height * self.capacity()
    self._drain_water_volume = 0 
    self._water_in = 0
    self._water_out = 0
    self._water_spill = 0

    # dynamic properties
    self._state = TankState.Static if self._water_volume else TankState.Empty

    # for printing
    self.ptoken = "state"

  #
  # state testing methods 
  #
  def isDraining( self ):
    return self.water_out() > self.water_in() 

  def isEmpty( self ):
    return self.water_volume() < 0.000001

  def isFilling( self ):
    return self.water_out() < self.water_in() 

  def isFull( self ):
    return self.water_volume() >= self.capacity() 

  def isOverflow( self ):
    return (self.isFull() and self.isFilling()) or (self.water_volume() > self.capacity())
  
  def isStatic( self ):
    return not self.isFull() and not self.isFilling()

  def update( self ):
    if self.isEmpty():
      self._state = TankState.Empty
    elif self.isOverflow():
      self._state = TankState.Overflow
    elif self.isFull():
      self._state = TankState.Full
    elif self.isFilling():
      self._state = TankState.Filling
    elif self.isDraining():
      self._state = TankState.Draining
    else:
      self._state = TankState.Static

  # 
  # integration method
  #
  def step( self ):
    # dummy
    drain_volume = self._drain.volume()

    # temp
    t_water_volume = self.water_volume() + self.water_in()

    # only set water volume is tank isn't full
    if not (self.state() == TankState.Full or self.state() == TankState.Overflow):
      self._water_volume = t_water_volume

    # only update water out if there is water
    if self.water_volume():
      # set water out
      drain_amt = drain_volume * self._drain.valve()
      self._water_out = drain_amt if drain_amt < self.water_volume() else self.water_volume()

      # lessen the water volume
      self._water_volume = self.water_volume() - self.water_out() 
      if self._water_volume < 0:
        self._water_volume = 0

      # make sure valves aren't < 0
      self._water_volume = self.water_volume() if self.water_volume() else 0

    # check for the tank overflowing
    if t_water_volume > self.capacity():
      self._water_spill = t_water_volume - self.capacity()
      self._water_volume = self.capacity()

    self.update()

  def println( self, what ):
    self.ptoken = what
    y = "Tank: " + what + ": " + self.__str__()
    return y

  def __str__( self ):
    y = None
    if self.ptoken == "state":
      y = TankState.translator[ self.state() ]
    elif self.ptoken == "water volume":
      y = self.water_volume()
    elif self.ptoken == "water height":
      y = self.water_height()
    elif self.ptoken == "water in":
      y = self.water_in()
    elif self.ptoken == "water out":
      y = self.water_out()
    elif self.ptoken == "valve":
      y = self.valve()
    return str(y)

from Constants import *

if __name__ == "__main__":
  import argparse

  parser = argparse.ArgumentParser()
  parser.add_argument( "n", help="number of runs", default=10, action="store_true" )
  parser.add_argument( "tank_radius", help="the radius of the tank", 
                        default=DEFAULT_LARGE_TANK_RADIUS, action="store_true" )
  parser.add_argument( "tank_height", help="the height of the tank", 
                        default=DEFAULT_LARGE_TANK_HEIGHT, action="store_true" )
  parser.add_argument( "drain_radius", help="the radius of the drain", 
                        default=DEFAULT_DRAIN_RADIUS, action="store_true" )
  parser.add_argument( "drain_height", help="the height of the drain", 
                        default=DEFAULT_DRAIN_HEIGHT, action="store_true" )
  parser.add_argument( "valve_ratio", help="the tightness of the drain valve", 
                        default=DEFAULT_DRAIN_HEIGHT, action="store_true" )
  parser.add_argument( "init_water_height", help="the initial volume of water", 
                        default=DEFAULT_LARGE_TANK_RADIUS/float(2), action="store_true" )
  args = parser.parse_args()

  info = TankInfo( args.tank_radius, args.tank_height, args.drain_radius, 
                   args.drain_height, args.valve_ratio, args.init_water_height )
  tank = TankController( info )
  

